src/libsyntax/parse/parser/pat.rs

```diff
@@ -12,80 +12,135 @@ use crate::ThinVec;
 
 use errors::{Applicability, DiagnosticBuilder};
 
+type Expected = Option<&'static str>;
+
 impl<'a> Parser<'a> {
     /// Parses a pattern.
-    pub fn parse_pat(
+    ///
+    /// Corresponds to `pat<no_top_alt>` in RFC 2535 and does not admit or-patterns
+    /// at the top level. Used when parsing the parameters of lambda expressions,
+    /// functions, function pointers, and `pat` macro fragments.
+    pub fn parse_pat(&mut self, expected: Expected) -> PResult<'a, P<Pat>> {
+        self.parse_pat_with_range_pat(true, expected)
+    }
+
+    // FIXME(or_patterns, Centril | dlrobertson):
+    // remove this and use `parse_top_pat` everywhere it is used instead.
+    pub(super) fn parse_top_pat_unpack(&mut self, gate_or: bool) -> PResult<'a, Vec<P<Pat>>> {
+        self.parse_top_pat(gate_or)
+            .map(|pat| pat.and_then(|pat| match pat.node {
+                PatKind::Or(pats) => pats,
+                node => vec![self.mk_pat(pat.span, node)],
+            }))
+    }
+
+    /// Entry point to the main pattern parser.
+    /// Corresponds to `top_pat` in RFC 2535 and allows or-pattern at the top level.
+    pub(super) fn parse_top_pat(&mut self, gate_or: bool) -> PResult<'a, P<Pat>> {
+        // Allow a '|' before the pats (RFCs 1925, 2530, and 2535).
+        if self.eat_or_separator() && gate_or {
+            self.sess.or_pattern_spans.borrow_mut().push(self.prev_span);
+        }
+
+        self.parse_pat_with_or(None, gate_or, true)
+    }
+
+    /// Parses a pattern, that may be a or-pattern (e.g. `Foo | Bar` in `Some(Foo | Bar)`).
+    /// Corresponds to `pat<allow_top_alt>` in RFC 2535.
+    fn parse_pat_with_or(
         &mut self,
-        expected: Option<&'static str>
+        expected: Expected,
+        gate_or: bool,
+        top_level: bool
     ) -> PResult<'a, P<Pat>> {
-        self.parse_pat_with_range_pat(true, expected)
+        // Parse the first pattern.
+        let first_pat = self.parse_pat(expected)?;
+        self.maybe_recover_unexpected_comma(first_pat.span, top_level)?;
+
+        // If the next token is not a `|`,
+        // this is not an or-pattern and we should exit here.
+        if !self.check(&token::BinOp(token::Or)) && self.token != token::OrOr {
```

Comment:
I guess you can look-ahead for the absence of various tokens like `, ] } )` and try to recover? I'd prefer to do more diagnostics polish in a follow up tho.

Reply:
I don't disagree, it's just that I noticed how hard it was to deal on expression parsing recovery because of similar things so I wanted to bring it up in case you might have some ideas on how to restrict it. Not critical for this PR.

Reply:
Thanks for the heads-up. :)

A cheap solution code-wise would be to reuse the generic list parsing infra and use `|` as the separator. Then you also need to check that `trailing == false` and error otherwise. (that's actually a good opportunity for recovery in case of a trailing `|`). The drawback to this is that it will unconditionally allocate a `Vec<P<Pat>>` which you will need to then convert back to `P<Pat>` in the case of a single element. That might pessimize parsing perf but it is probably negligible in the grand scheme of things... But let's revisit in a different PR. ^^